subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/IvyDependencyDescriptor.java

```diff
@@ -191,10 +191,14 @@ private void findMatches(ComponentIdentifier fromComponent, ComponentResolveMeta
 
     @Override
     public List<ExcludeMetadata> getConfigurationExcludes(Collection<String> configurations) {
-        List<ExcludeMetadata> rules = Lists.newArrayList();
+        if (excludes.isEmpty()) {
+            return Collections.emptyList();
+        }
+        Set<String> asSet = ImmutableSet.copyOf(configurations);
+        List<ExcludeMetadata> rules = Lists.newArrayListWithCapacity(excludes.size());
         for (Exclude exclude : excludes) {
             Set<String> ruleConfigurations = exclude.getConfigurations();
-            if (include(ruleConfigurations, configurations)) {
+            if (include(ruleConfigurations, asSet)) {
                 rules.add(exclude);
             }
         }
```

Comment:
Since the number of configurations is (normally) limited, I'd like to try building the `MultiMap<String, ExcludeMetadata>` once when constructing the `IvyDependencyDescriptor`. I guess I'm thinking we can get rid of `org.gradle.internal.component.model.Exclude` altogether, and just have a Map of configuration -> List<ExcludeMetadata>.

Reply:
We could do the same thing for `dependencyArtifacts`.